Example container images and instructions ...

castedo opened the pull request:
... for msprime development with nest containers on ubuntu 20.04

Enhancement #1

jeromekelleher commented on nests/README.md:
```diff
@@ -0,0 +1,71 @@
+can be run without root privilege.
+
+Loosely speaking, using `cnest` and `create-homey-nest`, enables workflows
+similar to 'virtualenv` and `conda`, but with near total isolation of the
```
Mismatched quote here on virtualenv, mixed back and forward ticks

jeromekelleher commented on nests/ubuntu-20.04-mspdev/development.txt:
```diff
@@ -0,0 +1,23 @@
+attrs
```
Do we need to have this here? This list of development dependencies is pretty fluid and it'll get out of date quickly. It'd be better to pip install the dev dependencies from the msprime/tskit repo directly I'd say.

castedo replied:
No need to have it here. I'm exploring and will commit a less hacky approach. Still a bit hacky, but less hacky.
There is also duplication between the list of required deb package names in the Dockerfile and the list of deb package names in the installation instructions. Might be worth having a single script file which contains the platform specific commands to run, have that file near msprime/requirements/*.txt, and then have both doc generation and this container building script read the same file to run the platform specific required system package installing lines.

castedo replied:
Commit in PR branch now, with this duplicated file deleted.
The Dockerfile is in theory less reproducible now. But the reproducibility of a container images from it's Docker file directory IMO is on a spectrum. I think having ~/src/msprime and a comment that it's a local git clone of the msprime github repository URL is reproducible enough for it's purpose.
That said, this might mean this doesn't work with `docker` even though `buildah` (podman) is fine with this. I stopped using docker when podman came out because docker, at least early versions of it, had a maniacal aversion to allowing functionality which would prevent reproducibility.

jeromekelleher reviewed (commented):
Looks good, thanks @castedo! Some comments above.

castedo commented:
What's the norm on when people click the "Re-request review"? Does this mean "all conversations resolved, I think the PR is ready to merge now" or does it mean "I request you spend some more time looking at new changes?"

Also, is it preferable that squashing happen after reviews (but before merge) rather than during the reviewing?

jeromekelleher commented:
> What's the norm on when people click the "Re-request review"? Does this mean "all conversations resolved, I think the PR is ready to merge now" or does it mean "I request you spend some more time looking at new changes?"

There isn't really an established convention for that here - I think we'd usually just ping each other on the comment thread when something is ready for another look.

> Also, is it preferable that squashing happen after reviews (but before merge) rather than during the reviewing?

It's best to do it at the end, I think, otherwise the comments on bits of the code become hard to follow. Doesn't really matter unless it's a big code drop.

agladstein commented:
Want me to test these out?

castedo commented:
@agladstein That would be nice, thx. What might be most useful is a 2nd opinion on the pros and cons of using this kind of container based workflow instead of conda or virtualenv. I'm a fan of containers so I'm probably a bit blind to the pros of conda/virtuanev and the cons of developing and testing in these "nest" containers.

I have only been using this on my RHEL8 laptop. It most likely works fine on Fedora, but I have not tested. Not sure how many people in the tskit-dev universe are developing on RHEL/CentOS/Fedora. I read a page recently that leads me to believe this "nest" container workflow/scripts are unlikely to ever be useful on macOS. In contrast the new Linux add-on for Windows might be a platform where this setup could be useful.
That all said, obviously macOS and Windows developer today can run Linux VMs on their non-Linux OS. And then they can approx do what someone on a Linux laptop/desktop can do. Used to do that on Macs for years.

castedo commented:
Oh, and if you want to be really bleeding edge, try these scripts with podman on debian bullseye. I read that podman is on it but I've got to imagine it's really buggy.

agladstein commented:
okay, cool. I should be able to get to it by this weekend.

castedo committed:
Example container images and instructions

  - updated to latest (incl pytest)
  - general container and C dev focusedcontainer
  - built for ubuntu 20.04

castedo commented:
What do people think of me updating these "nest" containers in this `tskit-dev/containers` repository vs a repository I've created: https://github.com/castedo/tskit-bud ?

I'm guessing it will be more convenient to have me merging my updates into main over in tskit-bud rather than giving me access to merging into main here ... or worse, somebody having to constantly review and merge by updates.

I don't get the sense that anybody else is using `podman` to run "personal" containers.

On the downside, maybe it's less likely people will find them. And I am keeping them up to date. They are the scripts/Dockerfiles that make the containers I use for all tskit related development.

castedo committed:
Add Debian (9) stretch circleci nest; remove Ubuntu 18.04

benjeffery commented:
> What do people think of me updating these "nest" containers in this `tskit-dev/containers` repository vs a repository I've created: https://github.com/castedo/tskit-bud ?

Quite happy to have them here and to merge this. @agladstein - did you get a chance to try this out in the end?

castedo commented:
@benjeffery Merging this here works for me. It's ready to go now. I did some mild squashing.

agladstein commented:
Sorry guys! Kept getting pushed down my to do list. I haven't gotten a chance to try them out yet. If you want to go ahead and pull, I can try them later and make an issue if there is any problem.

castedo commented:
No worries @agladstein. The usefulness of these "nest" containers I suspect is limited to those that are using `podman` which is limited to only a few distros at the moment. Net net I'm probably the only one who's going to be using them. :)
But they're here if any podman users want to do development within a container.
